src/parcsr_ls/par_mgr.c

```diff
@@ -3840,166 +4181,466 @@ HYPRE_Int hypre_blockRelax_solve (hypre_ParCSRMatrix *A,
+            }
+            if (++cnt == whole_num_points) { break; }
+         }
+         else
+         {
+            row_offset++;
+         }
+      }
+      // remaining points
+      for (i = whole_num_points; i < num_points; i++)
+      {
+         if (CF_marker[i] == point_type)
+         {
+            bidx = n_block;
+            ridx = cnt - whole_num_points;
+            bidxm1 = whole_num_points;
+            bidxp1 = num_points;
+            for (ii = A_diag_i[i]; ii < A_diag_i[i + 1]; ii++)
+            {
+               jj = A_diag_j[ii];
+               if (CF_marker[jj] == point_type)
+               {
+                  if (jj - row_offset >= bidxm1 && jj - row_offset < bidxp1 && fabs(A_diag_data[ii]) > HYPRE_REAL_MIN)
+                  {
+                     didx = bstart + ridx * left_size + jj - bidxm1 - row_offset;
+                     diag[didx] = A_diag_data[ii];
+                  }
+               }
+            }
+            cnt++;
+         }
+         else
+         {
+            row_offset++;
+         }
+      }
+   }
+   wall_time = time_getWallclockSeconds() - wall_time;
+
+   /*-----------------------------------------------------------------
+   * compute the inverses of all the diagonal sub-blocks
+   *-----------------------------------------------------------------*/
+   wall_time = time_getWallclockSeconds();
+   if (diag_type == 1) // compute the inverse
+   {
+      if (blk_size > 1)
+      {
+         hypre_BlockDiagInvLapack(diag, num_points, blk_size);
+      }
+      else
+      {
+         for (i = 0; i < num_points; i++)
+         {
+            if (fabs(diag[i]) < HYPRE_REAL_MIN)
+            {
+               diag[i] = 0.0;
+            }
+            else
+            {
+               diag[i] = 1.0 / diag[i];
+            }
+         }
+      }
+   }
+   wall_time = time_getWallclockSeconds() - wall_time;
+
+   *diag_ptr = diag;
+
+   return hypre_error_flag;
+}
 
-   hypre_Vector   *u_local = hypre_ParVectorLocalVector(u);
-   HYPRE_Real     *u_data  = hypre_VectorData(u_local);
+// Extract the block diagonal part of a A or a principal submatrix of A defined by a marker (point_type)
+// in an associated CF_marker array. The result is new block diagonal parCSR matrix.
+// If CF marker array is NULL, it returns the block diagonal of the original matrix A.
+// Options for diag_type are:
+// diag_type = 1: return the inverse of the block diagonals
+// otherwise : return the block diagonals
+// ***Input***
+// A - parCSR matrix
+// blk_size - Size of diagonal blocks to extract
+// CF_marker - Array prescribing submatrix from which to extract block diagonals. Ignored if NULL.
+// point_type - marker tag in CF_marker array to extract diagonal
+// diag_type - Type of block diagonal entries to return. Currently supports block diagonal or inverse block diagonal entries.
+// ***Output***
+// B_ptr: New block diagonal matrix
+//
+HYPRE_Int hypre_ParCSRMatrixBlockDiagMatrix(  hypre_ParCSRMatrix  *A,
```

Comment:
Same observation as for `hypre_ParCSRMatrixExtractBlockDiag`.

Reply:
This will be refactored too.